### PR TITLE
Restore uploading data + metadata

### LINKED ISSUE #95: Restore uploads
With new Generic Sensor interfaces and SensorManager we have the capture implemented via #85 .

Restore the zipping + UploadRequest logic in SensorManagerImpl.
Earlier it used to be in SensingEngineImpl

### PR DESCRIPTION
**Pull request by MJ1998**
Fixes #95 

More information in the issue.
This is simply restoring the UploadRequest record creation we used to do earlier in `SensingEngineImpl`. 

The current design is to be improved by introducing a default PostProcessor which would be responsible for all database consolidation (creation of CaptureInfo, ResourceInfo & UploadRequest records) - similar to [DefaultConsolidator in FhirEngine](https://github.com/google/android-fhir/blob/4742e60025df236abaca3884747f8429d68b504f/engine/src/main/java/com/google/android/fhir/sync/upload/ResourceConsolidator.kt#L42).

### CODE REVIEW

**Review by chandrashekar-s — APPROVED** [review]
Thanks for the changes @MJ1998 , they look fine

**Review by jingtang10 — CHANGES_REQUESTED** [review]
hey post checkup can we add some tests?

File: sensing/src/main/java/com/google/android/sensing/capture/CaptureUtil.kt [hunk 1]
```diff
@@ -16,43 +16,25 @@
...
-        toggleFlashFab.backgroundTintList =
-          ColorStateList.valueOf(ContextCompat.getColor(context, R.color.colorPrimary))
-      }
-    }
-
-    fun sensorsInvolved(captureType: CaptureType): List<InternalSensorType> {
-      return when (captureType) {
-        CaptureType.VIDEO_PPG -> listOf(InternalSensorType.CAMERA)
-        CaptureType.IMAGE -> listOf(InternalSensorType.CAMERA)
+    fun zipDirectory(sourceDirPath: Path, zipPath: Path) {
```
Comment (jingtang10) [inline sensing/src/main/java/com/google/android/sensing/capture/CaptureUtil.kt:28]: needs documentation

**Comment by MJ1998** [comment]
Yes I started them here: #99

File: sensing/src/main/java/com/google/android/sensing/impl/SensorManagerImpl.kt [hunk 2]
```diff
@@ -46,7 +50,11 @@ import kotlinx.coroutines.sync.withLock
 import kotlinx.coroutines.withContext
 import timber.log.Timber
 
-internal class SensorManagerImpl(context: Context, private val database: Database) : SensorManager {
+internal class SensorManagerImpl(
+  context: Context,
```
Comment (jingtang10) [inline sensing/src/main/java/com/google/android/sensing/impl/SensorManagerImpl.kt:54]: do you need this at all if this is not used anywhere else in this class?

File: sensing/src/main/java/com/google/android/sensing/capture/CaptureUtil.kt [hunk 3]
```diff
@@ -16,43 +16,25 @@
...
+      ZipOutputStream(BufferedOutputStream(FileOutputStream(zipPath.toFile()))).use { zipOut ->
+        Files.walk(sourceDirPath)
+          .filter { path -> !Files.isDirectory(path) } // Exclude directories from the initial walk
+          .forEach { path ->
+            val zipEntry = ZipEntry(sourceDirPath.relativize(path).toString())
```
Comment (jingtang10) [inline sensing/src/main/java/com/google/android/sensing/capture/CaptureUtil.kt:33]: i kinda like the idea of inlining this variable. but this is optional - try it and see if you think it's more succint. up to you!

File: sensing/src/main/java/com/google/android/sensing/impl/SensorManagerImpl.kt [hunk 4]
```diff
@@ -162,6 +170,8 @@ internal class SensorManagerImpl(context: Context, private val database: Databas
           CoroutineScope(Dispatchers.IO).launch {
             it.captureRequest?.let { request ->
               with(request) {
+                val toUploadFullUrl =
```
Comment (jingtang10) [inline sensing/src/main/java/com/google/android/sensing/impl/SensorManagerImpl.kt:173]: just inline this?

File: sensing/src/main/java/com/google/android/sensing/impl/SensorManagerImpl.kt [hunk 5]
```diff
@@ -170,30 +180,29 @@ internal class SensorManagerImpl(context: Context, private val database: Databas
...
+                     */
+                    withContext(lifecycleOwner.lifecycleScope.coroutineContext) {
+                      it.listener?.onComplete(captureInfo)
+                      /** TODO PostProcess. */
+                    }
+                    /** TODO move this to a default PostProcessor. */
+                    createUploadRequest(captureInfo, it.captureRequest!!)
                   }
-                } catch (e: ResourceNotFoundException) {
+                } catch (e: Exception) {
```
Comment (jingtang10) [inline sensing/src/main/java/com/google/android/sensing/impl/SensorManagerImpl.kt:200]: why this change? this is worse than before as it's less specific.

File: sensing/src/main/java/com/google/android/sensing/impl/SensorManagerImpl.kt [hunk 6]
```diff
@@ -227,6 +236,40 @@ internal class SensorManagerImpl(context: Context, private val database: Databas
     }
   }
 
+  /**
+   * Zips the folder data + metadata were captured to and add an UploadRequest in the database. TODO
```
Comment (jingtang10) [inline sensing/src/main/java/com/google/android/sensing/impl/SensorManagerImpl.kt:240]: is the documentation accurate? or maybe this function needs a better name?

Comment (jingtang10) [reply sensing/src/main/java/com/google/android/sensing/impl/SensorManagerImpl.kt:240]: also don't use `+`. sounds too casual.

Comment (jingtang10) [reply sensing/src/main/java/com/google/android/sensing/impl/SensorManagerImpl.kt:240]: actually this documentation makes me think this function needs to be split into two.

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: sensing/src/main/java/com/google/android/sensing/impl/SensorManagerImpl.kt (hunk 2 above)
Comment (MJ1998) [reply sensing/src/main/java/com/google/android/sensing/impl/SensorManagerImpl.kt:54]: I am thinking of merging SensingEngine APIs (CRUD APIs) into SensorManager to make it a unified interface. So 3 APIs will be moved:-
1. getCaptureInfo (captureId)
2. getCaptureInfo (externalIdentifier)
3. deleteDataInCapture(captureId)

Will need context in implementations of these APIs.

File: sensing/src/main/java/com/google/android/sensing/impl/SensorManagerImpl.kt (hunk 5 above)
Comment (MJ1998) [reply sensing/src/main/java/com/google/android/sensing/impl/SensorManagerImpl.kt:200]: Apart from ResourceNotFoundException, the createUploadRequest may throw other exceptions like IOException (when zipping) or Transaction error when inserting UploadRequest into the db.
So I thought of at least catching all if not the specific ones.
Can later shift to the specific exceptions when any application asks for it.

**Comment by MJ1998** [comment]
Seems like this got merged because of merging its child PR: #101 .
Raised [this issue](https://github.com/google-research/CVD-paper-mobile-camera-example/issues/105) to address comments here.
